### PR TITLE
Check that concentration parameters of Dirichlet distribution are all > 0

### PR DESCRIPTION
**Pull request by AlexAndorra**
As dicussed with @junpenglao, the Dirichlet distribution doesn't check that the concentration parameters (`a` vector) are all strictly positive when the distribution is initialized. 

Currently, the user can define `pm.Dirichlet("p", a=np.array([-1, 2, 3]), shape=3)` in a model and PyMC won't complain before the user tries to sample from the model (`BadInitialEnergy Error`).

As the concentration parameters *have* to be > 0, this PR just checks that it is the case before the user can do anything else.
Thanks in advance for the review, and I'm there for any comment/change. PyMCheers ✌️

### CODE REVIEW

**Comment by AlexAndorra** [comment]
I'm not sure I understand CodeCov output here: does this mean I have to modify something in all of the mentioned "Impacted files"?

**Comment by twiecki** [comment]
Does it work if the input is a RV?

**Comment by ColCarroll** [comment]
This looks like just a buggy report. I wouldn't worry.

**Comment by ColCarroll** [comment]
(note: thomas' question is still a good one)

**Comment by AlexAndorra** [comment]
Thanks @ColCarroll -- kind of a relief 😅 
@twiecki, do you mean if the `a` vector is a probability distribution?

**Comment by twiecki** [comment]
Yes, exactly.

**Comment by AlexAndorra** [comment]
The condition I added to check _strict_ positiveness is only for arrays -- so no; I should add this case. I suppose positiveness is already taken care of; I just have to figure out how to add the "strict" 😉

**Comment by AlexAndorra** [comment]
Still getting esoteric CodeCov warnings, but the tests passed 🎉 
Turns out the strict positiveness was a problem only when `a` was _not_ an RV, so I restricted the check to only those cases.
I'll also enabled the possibility to specify `a` as a list in addition to a numpy array, and I added some checks on `a`'s type.
Happy to hear what you guys think 🖖

**Comment by twiecki** [comment]
@AlexAndorra A test would be good.

**Comment by AlexAndorra** [comment]
Is think there already are tests, aren't there? In `test_distributions.py` and `test_distributions_random.py`

**Comment by twiecki** [comment]
@AlexAndorra Apparently not of passing in negative values, otherwise that test would have failed I think.

**Comment by AlexAndorra** [comment]
Ow ok -- sorry I'm not well versed in testing.
Very busy this week but I'll take a look ASAP!

**Comment by twiecki** [comment]
```
=================================== FAILURES ===================================
_________________ TestMatchesScipy.test_dirichlet_init_fail[3] _________________
self = <pymc3.tests.test_distributions.TestMatchesScipy object at 0x7f441e645c88>
n = 3
    @pytest.mark.parametrize('n', [3, 4])
    def test_dirichlet_init_fail(self, n):
        with Model():
            with pytest.raises(ValueError) as err:
                _ = Dirichlet('x', a=np.zeros(n), shape=n)
>           err.match("All concentration parameters (a) must be > 0.")
E           AssertionError: Pattern 'All concentration parameters (a) must be > 0.' does not match 'All concentration parameters (a) must be > 0.'
pymc3/tests/test_distributions.py:959: AssertionError
_________________ TestMatchesScipy.test_dirichlet_init_fail[4] _________________
self = <pymc3.tests.test_distributions.TestMatchesScipy object at 0x7f4415c37f28>
n = 4
    @pytest.mark.parametrize('n', [3, 4])
    def test_dirichlet_init_fail(self, n):
        with Model():
            with pytest.raises(ValueError) as err:
                _ = Dirichlet('x', a=np.zeros(n), shape=n)
>           err.match("All concentration parameters (a) must be > 0.")
E           AssertionError: Pattern 'All concentration parameters (a) must be > 0.' does not match 'All concentration parameters (a) must be > 0.'
pymc3/tests/test_distributions.py:959: AssertionError
```

File: pymc3/tests/test_distributions.py [hunk 1]
```diff
@@ -944,17 +944,39 @@ def test_lkj(self, x, eta, n, lp):
...
+
+    @pytest.mark.parametrize('n', [3, 4])
+    def test_dirichlet_init_fail(self, n):
+        with Model():
+            with pytest.raises(ValueError) as err:
+                _ = Dirichlet('x', a=np.zeros(n), shape=n)
+            err.match("All concentration parameters (a) must be > 0.")
+            with pytest.raises(ValueError) as err:
+                _ = Dirichlet('x', a=np.array([-1.] * n), shape=n)
+            err.match("All concentration parameters (a) must be > 0.")
```
Comment (lucianopaz) [inline pymc3/tests/test_distributions.py]:
```suggestion
            err.match(r"All concentration parameters \(a\) must be > 0.")
```

Parenthesis are a special character in regular expressions so you have to escape them with a backlash in the `match`

File: pymc3/tests/test_distributions.py [hunk 2]
```diff
@@ -944,17 +944,39 @@ def test_lkj(self, x, eta, n, lp):
...
+            {'a': Vector(Rplus, n)},
+            dirichlet_logpdf
+        )
+
+    @pytest.mark.parametrize('n', [3, 4])
+    def test_dirichlet_init_fail(self, n):
+        with Model():
+            with pytest.raises(ValueError) as err:
+                _ = Dirichlet('x', a=np.zeros(n), shape=n)
+            err.match("All concentration parameters (a) must be > 0.")
```
Comment (lucianopaz) [inline pymc3/tests/test_distributions.py]:
```suggestion
            err.match(r"All concentration parameters \(a\) must be > 0.")
```

Parenthesis are a special character in regular expressions so you have to escape them with a backlash in the `match`

File: pymc3/tests/test_distributions.py [hunk 3]
```diff
@@ -944,17 +944,39 @@ def test_lkj(self, x, eta, n, lp):
...
+            Dirichlet,
+            Simplex(n),
+            {'a': Vector(Rplus, n)},
+            dirichlet_logpdf
+        )
+
+    @pytest.mark.parametrize('n', [3, 4])
+    def test_dirichlet_init_fail(self, n):
+        with Model():
+            with pytest.raises(ValueError) as err:
```
Comment (lucianopaz) [inline pymc3/tests/test_distributions.py]: You can pass a `match` kwarg in the `pytest.raises` context manager and avoid storing it as `err` and later calling the `match` method. You can look at the [pytest docs](https://docs.pytest.org/en/latest/assert.html#assertions-about-expected-exceptions)

**Review by lucianopaz — CHANGES_REQUESTED** [review]

**Comment by AlexAndorra** [comment]
Thanks guys, this was very helpful!
Tests passed this time (notwithstanding the buggy codecov) 🍾

**Comment by twiecki** [comment]
Thanks!

**Pull request merged**